peregrine/resources/submission/__init__.py

```diff
@@ -133,12 +135,34 @@ def generate_schema_file(graphql_schema):
     with open(query_file, 'r') as f:
         query = f.read()
 
-    with open(schema_file, 'w') as f:
-        result = graphql_schema.execute(query)
-        data = {'data': result.data}
-        if result.errors:
-            data['errors'] = [err.message for err in result.errors]
-        json.dump(data, f)
+    try:
+        with open(schema_file, 'w') as f:
+            # lock file (prevents several processes from generating the schema at the same time)
+            fcntl.flock(f, fcntl.LOCK_EX | fcntl.LOCK_NB)
+
+            # generate the schema file
+            result = graphql_schema.execute(query)
+            data = {'data': result.data}
+            if result.errors:
+                data['errors'] = [err.message for err in result.errors]
+            json.dump(data, f)
+
+            fcntl.flock(f, fcntl.LOCK_UN) # unlock file
+    except:
+        # wait for file unlock (end of schema generation) before proceeding
+        timeout = time.time() + 60*5 # 5 minutes from now
+        while True:
+            try:
+                with open(schema_file, 'w') as f: # try to access+lock the file
+                    fcntl.flock(f, fcntl.LOCK_EX | fcntl.LOCK_NB)
+                    fcntl.flock(f, fcntl.LOCK_UN)
+                break # file is available -> schema has been generated -> process can proceed
+            except IOError: # file is still unavailable -> process waits
+                pass
+            if time.time() > timeout:
+                app_logger.warning('{} generation timeout: this process is proceeding without waiting for file generation.'.format(schema_file))
+                break
```

Comment:
I think it's cool to aggressively log what peregrine is doing at startup - it can be frustrating waiting for the service to get healthy, "well - I guess it's downloading the dictionary or something ..."

Reply:
```
spawned uWSGI master process (pid: 1384)
spawned uWSGI worker 1 (pid: 1385, cores: 1)
spawned uWSGI worker 2 (pid: 1386, cores: 1)
Process 1 is generating the graphql schema file schema.json.
Process 2 is waiting for schema.json generation.
Process 1 is done generating schema.json.
WSGI app 0 (mountpoint='') ready in 28 seconds on interpreter 0x1d64ea0 pid: 1386 (default app)
WSGI app 0 (mountpoint='') ready in 28 seconds on interpreter 0x1d64ea0 pid: 1385 (default app)
```